### ISSUE TITLE
Microsoft Excel is waiting for another application to complete an OLE action

### ISSUE THREAD

**Issue by martinoclaudio** [florentbr/SeleniumBasic#61]
Hello Florent,
here I am with a new issue that may be related to VBA and Selenium. 
I am running Windows 7 with Excel 2010 , Firefox 42.0 and SeleniumBasic2.0.5

I am running at the same time 3 Excel instances and every instance ist using two Drivers defined as Driver and DriverB, please see below:
````
Public driver As New Selenium.FirefoxDriver
Public By As New By, Assert As New Selenium.Assert, Verify As New Selenium.Verify, Waiter As New Selenium.Waiter
Public Utils As New Selenium.Utils
Public keys As New Selenium.keys
Public driverB As New Selenium.FirefoxDriver
Public ByB As New By, AssertB As New Selenium.Assert, VerifyB As New Selenium.Verify, WaiterB As New Selenium.Waiter
Public UtilsB As New Selenium.Utils
Public keysB As New Selenium.keys
````
From the Timings I am setting:
````
    driver.Timeouts.ImplicitWait = 3000
    driver.Timeouts.Script = 3000
    driver.Timeouts.Server = 60000
    driver.Timeouts.PageLoad = 60000
    driverB.Timeouts.ImplicitWait = 3000
    driverB.Timeouts.Script = 3000
    driverB.Timeouts.Server = 60000
    driverB.Timeouts.PageLoad = 60000
````
This OLE error is happening only once per 1 Million starts of my Selenium Routine. However I tried everything that I thought could help - without success so far. 
Now I am wonding if it is in general possible and stable to run Driver and DriverB in parallel at the same Excel instance.
Would be great again to get some help or ideas from you!

Thanks
Martin

**Comment by florentbr** [comment]
Can you provide a concise and reproducible example?
How do you run/launch the code?
Parallel execution is only possible with multiple instances of Excel. Unless I'm wrong, drivers on the same instance can only be executed sequentially.

**Comment by martinoclaudio** [comment]
The Problem is it happening very rarely. I am starting the routines by Excel Events.
Currently I am suspecting the following:
Within Excel/VBA the OLERequestPendingTimeout is set to 5000ms and cannot be changed.
I am setting the Driver.Timeouts.Server to 60000ms. 
Probably the Selenium COM library is not responding within the OLERequestPendingTimeout Limit and as I set the  Driver.Timeouts.Server to a higher time there is no error generated by the Selenium Server.
Furthermore the error is only happening when I am switching between Driver and DriverB.
As a trial I am currently changing the program to only one Driver instance and secondly I am reducing the Driver.Timeouts.Server to 4500ms.
Do you know the default setting for Driver.Timeouts.Server ?

I will keep you informed if my actions will solve the Problem.

Thanks again
Martin

**Comment by florentbr** [comment]
The timeout for the server is set to 60000ms by default.
You can try to:
* Use the last version of the driver (the waiting mechanism has been updated since v2.0.5).
* Use only the driver.Wait method and not the Sleep from kernell32 which is highly io blocking
* Set Application.DisplayAlerts = False when executing the script
* Set Application.EnableEvents = False when writing in a spreadsheet

It could also be a deadlock issue from the two drivers competing for the same resource.

**Comment by martinoclaudio** [comment]
Point 3 and 4 were already integrated.

By the way one could switch off this Excel OLE alert also by the following:
````
Option Explicit 
Private Declare Function CoRegisterMessageFilter Lib "OLE32.DLL" (ByVal lFilterIn As Long, ByRef lPreviousFilter) As Long' - See more at: http://msgroups.net/excel/excel-is-waiting-for-another-application-to-complete-ol/266667#sthash.yqrWFLHB.dpuf
'Hide OLE blocked process message box
Dim lMsgFilter As Long 
CoRegisterMessageFilter 0&, lMsgFilter 
````
However this would be my last step.

Currently I also believe that this is the deadlock issue from the two drivers competing for the same resource what you mentioned.
I never have had this issue when using only one Driver in one Excel instance with similar programs.

From next week I will test my new version with only one Driver per Excel instance, but still with up to 6 instances running in parallel.

Thanks again
Martin

**Comment by uzhas-sovka** [comment]
Hello Florent, maybe it's a wrong place to ask questions. How can I keep browser running after VBA script is complete? Thank you.

**Comment by florentbr** [comment]
@uzhas-sovka, your question is no related to this issue. You'll have to create a new one to expect some support.

**Comment by martinoclaudio** [comment]
Hello Florent,
after running in every instance only one Driver I haven't faced any problems with the OLE anymore.
However I also changed what you recommended and updated to Version 2.07 and changed the Sleep from kernell32 to Driver.wait from Selenium.

I managed to run up to 12 independet Excel instances running Selenium without problems. I also can confirm that with Driver.quit the temporary Selenium Files are deleted in the right way.

Just a small issue I saw is that Driver.url is giving the actual URL in the active Firefox tab, but Driver.baseurl is giving that one of the last opened tab.
I am not sure whether this is an issue at all - so I just want to mention it.

Thanks again for your great Support!

**Comment by florentbr** [comment]
The .BaseUrl property holds the root of the Url used with .Get.
It makes it possible to call .Get with a short URL:

```vb
Dim driver As New FirefoxDriver

driver.Get "https://www.google.co.uk"
Debug.Assert 0

driver.Get "/intl/en/about"
Debug.Assert 0

driver.baseUrl = "https://www.google.co.uk/intl/en"
driver.Get "/policies/privacy"
Debug.Assert 0

driver.Quit
```
